### PR TITLE
Separate service id from task

### PR DESCRIPTION
**Pull request by pbetkier**
Consul service used to be registered under the same id as its related marathon task id. This PR separates these two ids. Marathon task id is remembered in service's `marathon-task` tag. Also did some refactorings: extracted Service and ServiceId domain types, and also ServiceRegistry interface to make things more clear by hiding Consul internals.

I'm still working on this PR, going to cleanup some code and verify that all branches are covered with tests. But the main idea is there, so waiting for your feedback.

It's a prerequisite to #89, because we would register more than one service for a given marathon task.

### CODE REVIEW

**Comment by janisz** [comment]
I think this should be split into 2 PR: introduce Service and ServiceId domain types and change ServiceId to generated uuid. This will make review and testing easier. Another thing changing IDs could be harmful so we need to test impact of this action.

**Comment by pbetkier** [comment]
Even if do we split, we have to test this PR thoroughly because it's a significant change. Why not do it once? I've split the change into two separate commits for the review's sake, but I don't see the point in splitting it into separate PRs. One would contain the feature, but introducing complexity in the codebase and the other would only clean things up. Still, we merge either both or nothing IMO :)

**Comment by janisz** [comment]
I thought we can merge refactoring first and then ID change. With refactoring most of tests should remain untouched and by this we will ensure change doesn't break anything.

**Comment by janisz** [comment]
I created dedicated task (#100) for extracting domain service object. It should be backward comaptible and we could release it as 4.1 build with Go 1.7.

File: service/service.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package service
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/allegro/marathon-consul/apps"
+	"fmt"
```
Comment (janisz) [inline service/service.go]: `find ./ -name '*.go' | xargs goimports -w`

File: service/service.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+type Service struct {
+	ID                      ServiceId
+	Name                    string
+	Tags                    []string
+	RegisteringAgentAddress string
+}
+
+func (s *Service) TaskId() (apps.TaskId, error) {
```
Comment (janisz) [inline service/service.go]: Why pointer receiver?

File: consul/consul.go [hunk 3]
```diff
@@ -63,58 +56,71 @@ func (c *Consul) getServicesUsingProviderWithRetriesOnAgentFailure(provide Servi
...
+				for _, consulServiceInstance := range consulServiceInstances {
+					allInstances = append(allInstances, c.consulServiceToService(consulServiceInstance))
+				}
 			}
 		}
 	}
 	return allInstances, nil
 }
 
+func (c *Consul) consulServiceToService(consulService *consulapi.CatalogService) *service.Service {
```
Comment (janisz) [inline consul/consul.go]: How about moving this function to `service` package and make it constructor function e.g. `func FromConsul(consulService *consulapi.CatalogService) *service.Service`

File: consul/consul.go [hunk 4]
```diff
@@ -1,32 +1,25 @@
...
 
 import (
+	"errors"
 	"fmt"
 	"net/url"
 	"strings"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/allegro/marathon-consul/apps"
+	"github.com/allegro/marathon-consul/service"
```
Comment (janisz) [inline consul/consul.go]: Consul package shouldn't have dependency to service.

File: consul/consul.go [hunk 5]
```diff
@@ -162,9 +168,32 @@ func (c *Consul) register(service *consulapi.AgentServiceRegistration) error {
 	return err
 }
 
-func (c *Consul) Deregister(serviceId apps.TaskId, agentAddress string) error {
+func (c *Consul) DeregisterByTask(taskToDeregister apps.TaskId, agentAddress string) error {
```
Comment (janisz) [inline consul/consul.go]: `taskToDeregister` → `taskIDToDeregister` or just `taskID`

File: consul/consul.go (hunk 3 above)
Comment (pbetkier) [inline consul/consul.go]: Then service package would depend on consul package, not cool :)

File: consul/consul.go (hunk 4 above)
Comment (pbetkier) [inline consul/consul.go]: I'd say right the opposite, you don't want our domain types with business logic depend on infrastructure, external classes like consul.

Comment (janisz) [inline consul/consul.go]: Good point

File: consul/consul.go (hunk 3 above)
Comment (janisz) [inline consul/consul.go]: OK. But this could be a pure function not depended on `*Consul`

File: consul/consul.go [hunk 6]
```diff
@@ -35,24 +28,24 @@ func New(config ConsulConfig) *Consul {
...
+func (c *Consul) getServicesUsingProviderWithRetriesOnAgentFailure(provide ServicesProvider) ([]*service.Service, error) {
 	for retry := uint32(0); retry <= c.config.RequestRetries; retry++ {
 		agent, err := c.agents.GetAnyAgent()
 		if err != nil {
 			return nil, err
 		}
 		if services, err := provide(agent.Client); err != nil {
 			log.WithError(err).WithField("Address", agent.IP).
-				Error("An error occurred getting services from Consul, retrying with another agent")
+			Error("An error occurred getting services from Consul, retrying with another agent")
```
Comment (medzin) [inline consul/consul.go]: Are you sure it is Error and not a Debug/Notice message?

File: consul/consul.go [hunk 7]
```diff
@@ -162,9 +168,32 @@ func (c *Consul) register(service *consulapi.AgentServiceRegistration) error {
...
 }
 
-func (c *Consul) Deregister(serviceId apps.TaskId, agentAddress string) error {
+func (c *Consul) DeregisterByTask(taskToDeregister apps.TaskId, agentAddress string) error {
+	service, err := c.findServiceByTaskId(taskToDeregister)
+	if err != nil {
+		return err
+	}
+	c.Deregister(service)
+	return err
```
Comment (medzin) [inline consul/consul.go]: IMHO - better to be explicit here so`return nil`

File: consul/consul.go (hunk 3 above)
Comment (pbetkier) [inline consul/consul.go]: Good point, I'll make it a pure function.

File: consul/consul.go (hunk 6 above)
Comment (pbetkier) [inline consul/consul.go]: @janisz what do you think? Warn / Debug? Should we change levels in this PR?

Comment (janisz) [inline consul/consul.go]: I'll leave it as is (without changing indentation)

**Comment by pbetkier** [comment]
@janisz @dankraw I fixed according to review comments, can I merge this into develop?

**Comment by dankraw** [comment]
late but 👍  :>

**Pull request merged**